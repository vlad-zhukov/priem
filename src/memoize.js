import {Cache, CacheItem} from './Cache';
import {noop, type} from './helpers';

/**
 * @function isSameValueZero
 *
 * @description
 * are the objects equal based on SameValueZero
 *
 * @param {any} object1 the first object to compare
 * @param {any} object2 the second object to compare
 * @returns {boolean} are the two objects equal
 */
function isSameValueZero(object1, object2) {
    // eslint-disable-next-line no-self-compare
    return object1 === object2 || (object1 !== object1 && object2 !== object2);
}

/**
 * @function areKeysEqual
 *
 * @description
 * are the keys shallowly equal to one another
 *
 * @param {Array<any>} keys1 the keys array to test against
 * @param {Array<any>} keys2 the keys array to test
 * @returns {boolean} are the keys shallowly equal
 */
export function areKeysEqual(keys1, keys2) {
    if (keys1.length !== keys2.length) {
        return false;
    }

    for (let index = 0; index < keys1.length; index++) {
        if (!isSameValueZero(keys1[index], keys2[index])) {
            return false;
        }
    }

    return true;
}

export const PENDING = 0;
export const FULFILLED = 1;
export const REJECTED = 2;

function createTimeout(cache, item, maxAge, onCacheChange) {
    // eslint-disable-next-line no-restricted-globals
    const normalizeMaxAge = type(maxAge) === 'number' && isFinite(maxAge) ? maxAge : null;
    if (normalizeMaxAge !== null) {
        clearTimeout(item.timeoutId);
        // eslint-disable-next-line no-param-reassign
        item.timeoutId = setTimeout(() => {
            onCacheChange({args: item.key, forceRefresh: true});
        }, normalizeMaxAge);
    }
}

export default function memoize({fn, initialCache = [], maxSize = 1, maxAge = Infinity, onCacheChange = noop}) {
    const cache = Cache.fromArray(initialCache);

    function memoized(args, options) {
        let item = cache.findBy(cacheItem => areKeysEqual(cacheItem.key, args));
        let shouldRefresh = false;

        if (item === null) {
            if (cache.size >= maxSize) {
                const itemToRemove = cache.tail;
                cache.remove(itemToRemove);
                itemToRemove.destroy();
            }

            item = new CacheItem(args, {status: PENDING, value: null, reason: null});
            cache.prepend(item);
            shouldRefresh = true;
        } else {
            if (item !== cache.head) {
                cache.remove(item);
                cache.prepend(item);
            }
            const forceRefresh = (options && options.forceRefresh) || false;
            if (forceRefresh === true) {
                shouldRefresh = true;
            }
        }

        if (shouldRefresh) {
            createTimeout(cache, item, maxAge, onCacheChange);
            const itemValue = Object.assign(item.value, {status: PENDING, reason: null});
            itemValue.promise = fn
                .apply(this, args)
                .then(value => {
                    Object.assign(itemValue, {status: FULFILLED, value, reason: null});
                    onCacheChange({args});
                })
                .catch(error => {
                    Object.assign(itemValue, {status: REJECTED, value: null, reason: error});
                    onCacheChange({args});
                });
        }

        // console.log(Date.now(), cache.head.key, cache.head.value);

        return item.value;
    }

    Object.defineProperties(memoized, {
        cache: {
            value: cache,
        },
        isMemoized: {
            value: true,
        },
    });

    return memoized;
}
