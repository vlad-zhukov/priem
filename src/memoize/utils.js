export const createAreKeysEqual = isEqual =>
    /**
     * @function areKeysEqual
     *
     * @description
     * are the keys shallowly equal to one another
     *
     * @param {Array<any>} keys1 the keys array to test against
     * @param {Array<any>} keys2 the keys array to test
     * @returns {boolean} are the keys shallowly equal
     */
    (keys1, keys2) => {
        if (keys1.length !== keys2.length) {
            return false;
        }

        for (let index = 0; index < keys1.length; index++) {
            if (!isEqual(keys1[index], keys2[index])) {
                return false;
            }
        }

        return true;
    };

export const createGetKeyIndex = isEqual => {
    const areKeysEqual = createAreKeysEqual(isEqual);

    /**
     * @function getKeyIndex
     *
     * @description
     * get the index of the matching key
     *
     * @param {Array<Array<any>>} allKeys the list of all available keys
     * @param {Array<any>} keysToMatch the key to try to match
     *
     * @returns {number} the index of the matching key value, or -1
     */
    return (allKeys, keysToMatch) => {
        for (let index = 0; index < allKeys.length; index++) {
            if (areKeysEqual(allKeys[index], keysToMatch)) {
                return index;
            }
        }
        return -1;
    };
};

/**
 * @function isSameValueZero
 *
 * @description
 * are the objects equal based on SameValueZero
 *
 * @param {any} object1 the first object to compare
 * @param {any} object2 the second object to compare
 * @returns {boolean} are the two objects equal
 */
export const isSameValueZero = (object1, object2) =>
    // eslint-disable-next-line no-self-compare
    object1 === object2 || (object1 !== object1 && object2 !== object2);

/**
 * @function orderByLru
 *
 * @description
 * order the array based on a Least-Recently-Used basis
 *
 * @param {Array<any>} array the array to order
 * @param {any} value the value to assign at the beginning of the array
 * @param {number} startingIndex the index of the item to move to the front
 */
export const orderByLru = (array, value, startingIndex) => {
    let index = startingIndex;
    while (index--) { // eslint-disable-line no-plusplus
        array[index + 1] = array[index]; // eslint-disable-line no-param-reassign
    }
    array[0] = value;  // eslint-disable-line no-param-reassign
};
