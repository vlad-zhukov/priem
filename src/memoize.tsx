import {Cache, CacheItem, SerializableCacheItem, reduce} from './Cache';
import {type, isBrowser, noop, areKeysEqual} from './utils';

const DEFAULT_THROTTLE_MS = 150;

export enum STATUS {
    PENDING = 0,
    FULFILLED = 1,
    REJECTED = 2,
}

export type MemoizedKey = unknown[];

export type MemoizedValue = {
    status: STATUS;
    data: unknown;
    reason: Error | null;
    promise?: Promise<void>;
};

export type MemoizedCache = Cache<MemoizedKey, MemoizedValue>;

export type MemoizedCacheItem = CacheItem<MemoizedKey, MemoizedValue>;

export type MemoizedSerializableCacheItem = SerializableCacheItem<MemoizedKey, MemoizedValue>;

export function toSerializableArray(
    cache: MemoizedCache,
    filterFulfilled: boolean = false
): MemoizedSerializableCacheItem[] {
    return reduce<MemoizedSerializableCacheItem[], MemoizedKey, MemoizedValue>(cache, [], (acc, item) => {
        const {status, data, reason} = item.value;
        if (filterFulfilled !== true || status === STATUS.FULFILLED) {
            acc.push({
                key: item.key,
                value: {status, data, reason},
            });
        }
        return acc;
    });
}

export type CacheChangeCallback = (key: MemoizedKey, forceUpdate: boolean) => void;

function createTimeout(
    cache: MemoizedCache,
    item: MemoizedCacheItem,
    maxAge: number | null,
    onCacheChange: CacheChangeCallback
): void {
    if (isBrowser === true && maxAge !== null) {
        window.clearTimeout(item.expireId);
        item.expireId = window.setTimeout(() => {
            onCacheChange(item.key, true);
        }, maxAge);
    }
}

type MemoizeOptions<Args extends MemoizedKey> = {
    fn: (...args: Args) => Promise<unknown>;
    initialCache?: (MemoizedSerializableCacheItem | MemoizedCacheItem)[];
    maxSize?: number;
    maxAge?: number;
    onCacheChange?: CacheChangeCallback;
};

export type MemoizedFunction = {
    (args: MemoizedKey, forceRefresh?: boolean): MemoizedValue;
    cache: MemoizedCache;
    has: (args: MemoizedKey) => boolean;
};

export default function memoize<Args extends MemoizedKey>({
    fn,
    initialCache = [],
    maxSize = 1,
    maxAge = Infinity,
    onCacheChange = noop,
}: MemoizeOptions<Args>): MemoizedFunction {
    const cache: MemoizedCache = new Cache(initialCache);
    const normalizeMaxAge = type(maxAge) === 'number' && isFinite(maxAge) ? maxAge : null;

    function memoized(args: Args, forceRefresh: boolean = false): MemoizedValue {
        let item: MemoizedCacheItem | null = cache.findBy(cacheItem => areKeysEqual(cacheItem.key, args));
        let shouldRefresh = false;

        if (item === null) {
            if (cache.size >= maxSize) {
                const itemToRemove = cache.tail;
                cache.remove(itemToRemove);
                itemToRemove!.destroy();
            }

            item = new CacheItem(args, {status: STATUS.PENDING, data: null, reason: null});
            cache.prepend(item);
            shouldRefresh = true;
        } else {
            if (item !== cache.head && cache.remove(item) !== null) {
                cache.prepend(item);
            }

            if (forceRefresh === true) {
                shouldRefresh = true;
            }
        }

        if (shouldRefresh) {
            // Throttle refreshes
            const now = Date.now();
            const lastRefreshAt = item.lastRefreshAt || 0;
            if (now - lastRefreshAt > DEFAULT_THROTTLE_MS) {
                item.lastRefreshAt = now;
                createTimeout(cache, item, normalizeMaxAge, onCacheChange);
                const itemValue = Object.assign(item.value, {status: STATUS.PENDING, reason: null});
                itemValue.promise = fn(...args)
                    .then(data => {
                        Object.assign(itemValue, {status: STATUS.FULFILLED, data, reason: null});
                        onCacheChange(args, false);
                    })
                    .catch(error => {
                        Object.assign(itemValue, {status: STATUS.REJECTED, data: null, reason: error});
                        onCacheChange(args, false);
                    });
            }
        }

        return item.value;
    }

    Object.defineProperties(memoized, {
        cache: {
            value: cache,
        },
        has: {
            value: function has(args: MemoizedKey) {
                return !!cache.findBy(cacheItem => areKeysEqual(cacheItem.key, args));
            },
        },
    });

    // @ts-ignore
    return memoized;
}
