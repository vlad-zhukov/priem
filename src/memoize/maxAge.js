import {createGetKeyIndex, noop} from './utils';

/**
 * @private
 *
 * @function findExpirationIndex
 *
 * @description
 * find the index of the expiration based on the key
 *
 * @param {Array<Expiration>} expirations the list of expirations
 * @param {Array<any>} key the key to match
 * @returns {number} the index of the expiration
 */
export const findExpirationIndex = (expirations, key) => {
    for (let index = 0; index < expirations.length; index++) {
        if (expirations[index].key === key) {
            return index;
        }
    }
    return -1;
};

/**
 * @private
 *
 * @function clearExpiration
 *
 * @description
 * clear an active expiration and remove it from the list if applicable
 *
 * @param {Array<Expiration>} expirations the list of expirations
 * @param {any} expirationIndex the key to clear
 * @param {boolean} [shouldRemove] should the expiration be removed from the list
 */
export const clearExpiration = (expirations, expirationIndex, shouldRemove) => {
    if (~expirationIndex) {
        clearTimeout(expirations[expirationIndex].timeoutId);

        if (shouldRemove) {
            expirations.splice(expirationIndex, 1);
        }
    }
};

export const createOnCacheAddSetExpiration = (expirations, options, isEqual) => {
    const {maxAge, onCacheChange, onExpire} = options;

    const findKeyIndex = createGetKeyIndex(isEqual);

    /**
     * @private
     *
     * @function onCacheAdd
     *
     * @description
     * when an item is added to the cache, add an expiration for it
     *
     * @modifies {expirations}
     *
     * @param {Cache} cache the cache of the memoized function
     * @param {Options} moizedOptions the options passed to the memoized function
     * @param {function} moized the memoized function
     * @returns {void}
     */
    return function onCacheAdd(cache, moizedOptions, moized) {
        const key = cache.keys[0];

        if (!~findExpirationIndex(expirations, key)) {
            const expirationMethod = () => {
                const keyIndex = findKeyIndex(cache.keys, key);
                const value = cache.values[keyIndex];

                if (~keyIndex) {
                    cache.keys.splice(keyIndex, 1);
                    cache.values.splice(keyIndex, 1);

                    onCacheChange(cache, moizedOptions, moized);
                }

                const expirationIndex = findExpirationIndex(expirations, key);
                clearExpiration(expirations, expirationIndex, true);

                if (onExpire(key) === false) {
                    cache.keys.unshift(key);
                    cache.values.unshift(value);

                    createOnCacheAddSetExpiration(expirations, options, isEqual)(cache, moizedOptions, moized);
                    onCacheChange(cache, moizedOptions, moized);
                }
            };

            expirations.push({
                expirationMethod,
                key,
                timeoutId: setTimeout(expirationMethod, maxAge),
            });
        }
    };
};

export const createOnCacheHitResetExpiration = (expirations, options) => {
    const {maxAge} = options;

    /**
     * @private
     *
     * @function onCacheHit
     *
     * @description
     * when a cache item is hit, reset the expiration
     *
     * @modifies {expirations}
     *
     * @param {Cache} cache the cache of the memoized function
     * @returns {void}
     */
    return function onCacheHit(cache) {
        const key = cache.keys[0];
        const expirationIndex = findExpirationIndex(expirations, key);

        if (~expirationIndex) {
            clearExpiration(expirations, expirationIndex, false);
            // eslint-disable-next-line no-param-reassign
            expirations[expirationIndex].timeoutId = setTimeout(expirations[expirationIndex].expirationMethod, maxAge);
        }
    };
};

/**
 * @private
 *
 * @function getMaxAgeOptions
 *
 * @description
 * get the micro-memoize options specific to the maxAge option
 *
 * @param {Array<Expiration>} expirations the expirations for the memoized function
 * @param {Options} options the options passed to the moizer
 * @param {function} isEqual the function to test equality of the key on a per-argument basis
 * @returns {Object} the object of options based on the entries passed
 */
export const getMaxAgeOptions = (expirations, options, isEqual) => {
    const {maxAge, updateExpire} = options;

    let onCacheAdd = noop;
    // eslint-disable-next-line no-restricted-globals
    if (typeof maxAge === 'number' && isFinite(maxAge)) {
        onCacheAdd = createOnCacheAddSetExpiration(expirations, options, isEqual);
    }

    let onCacheHit = noop;
    if (onCacheAdd && updateExpire) {
        onCacheHit = createOnCacheHitResetExpiration(expirations, options);
    }

    return {onCacheAdd, onCacheHit};
};
