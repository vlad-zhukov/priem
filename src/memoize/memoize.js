import {noop, createGetKeyIndex, isSameValueZero, orderByLru} from './utils';
import {getMaxAgeOptions} from './maxAge';

export const PENDING = 0;
export const FULFILLED = 1;
export const REJECTED = 2;

/**
 * @function memoize
 *
 * @description
 * get the memoized version of the method passed
 *
 * @param {function} fn the method to memoize
 * @param {Object} [options={}] the options to build the memoizer with
 * @param {boolean} [options.isEqual=isSameValueZero] the method to compare equality of keys with
 * @param {number} [options.maxSize=1] the number of items to store in cache
 * @returns {function} the memoized method
 */
export default function memoize(fn, options = {}) {
    const {
        isEqual = isSameValueZero,
        maxSize = 1,
        maxAge = Infinity,
        onCacheChange = noop,
        onExpire = noop,
        updateExpire = false,
        ...extraOptions
    } = options;

    const getKeyIndex = createGetKeyIndex(isEqual);
    const expirations = [];
    const {onCacheAdd, onCacheHit} = getMaxAgeOptions(
        expirations,
        {onCacheChange, maxAge, onExpire, updateExpire},
        isEqual
    );

    const normalizedOptions = Object.assign({}, extraOptions, {
        isEqual,
        maxSize,
        maxAge,
        onCacheAdd,
        onCacheChange,
        onCacheHit,
        onExpire,
        updateExpire,
    });

    const keys = [];
    const values = [];
    const cache = {
        keys,
        values,
        get size() {
            return keys.length;
        },
    };

    /**
     * @function memoized
     *
     * @description
     * the memoized version of the method passed
     *
     * @param {...Array<any>} args the arguments passed, which create a unique cache key
     * @returns {any} the value of the method called with the arguments
     */
    function memoized(...args) {
        const keyIndex = getKeyIndex(keys, args);

        if (~keyIndex) {
            onCacheHit(cache, normalizedOptions, memoized);

            if (keyIndex) {
                orderByLru(keys, keys[keyIndex], keyIndex);
                orderByLru(values, values[keyIndex], keyIndex);

                onCacheChange(cache, normalizedOptions, memoized);
            }
        } else {
            if (keys.length >= maxSize) {
                keys.pop();
                values.pop();
            }

            const item = {
                status: PENDING,
                value: null,
                reason: null,
            };

            orderByLru(keys, args, keys.length);
            orderByLru(values, item, values.length);

            item.promise = fn
                .apply(this, args)
                .then(result => {
                    item.status = FULFILLED;
                    item.value = result;
                    item.reason = null;

                    onCacheHit(cache, normalizedOptions, memoized);
                    onCacheChange(cache, normalizedOptions, memoized);

                    return result;
                })
                .catch(error => {
                    item.status = REJECTED;
                    item.value = null;
                    item.reason = error;

                    onCacheHit(cache, normalizedOptions, memoized);
                    onCacheChange(cache, normalizedOptions, memoized);
                });

            onCacheAdd(cache, normalizedOptions, memoized);
            onCacheChange(cache, normalizedOptions, memoized);
        }

        return values[0];
    }

    Object.defineProperties(memoized, {
        cache: {
            configurable: true,
            get() {
                return cache;
            },
        },
        isMemoized: {
            configurable: true,
            get() {
                return true;
            },
        },
        options: {
            configurable: true,
            get() {
                return normalizedOptions;
            },
        },
    });

    return memoized;
}
