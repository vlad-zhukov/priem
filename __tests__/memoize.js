import delay from 'delay';
import memoize from '../src/memoize';

it('should memoize promises', async () => {
    const memoized = memoize(name => delay(200, {value: `Hello ${name}!`}), {maxSize: 2});

    expect(memoized('world')).toMatchSnapshot();

    await delay(300);
    expect(memoized('world')).toMatchSnapshot();
    expect(memoized('world')).toMatchSnapshot();
    expect(memoized('SpongeBob')).toMatchSnapshot();

    await delay(300);
    expect(memoized('SpongeBob')).toMatchSnapshot();
    expect(memoized('world')).toMatchSnapshot();
});

it('should not throw on rejected promises', async () => {
    const memoized = memoize(name => delay.reject(200, {value: new Error(`Hello ${name}!`)}), {maxSize: 2});

    expect(memoized('world')).toMatchSnapshot();

    await delay(300);
    expect(memoized('world')).toMatchSnapshot();
    expect(memoized('world')).toMatchSnapshot();
    expect(memoized('SpongeBob')).toMatchSnapshot();

    await delay(300);
    expect(memoized('SpongeBob')).toMatchSnapshot();
});

it('should export the internal cache', async () => {
    const memoized = memoize(name => delay(200, {value: `Hello ${name}!`}));

    expect(memoized.cache).toMatchSnapshot();
    expect(memoized.cache.size).toBe(0);

    memoized('world');
    expect(memoized.cache).toMatchSnapshot();

    await delay(300);
    expect(memoized.cache).toMatchSnapshot();
    expect(memoized.cache.size).toBe(1);
});

it('should default `maxSize` to 1', async () => {
    const memoized = memoize((name1, name2) =>
        delay(200, {
            value: `Hello ${name1}${name2 ? `and ${name2}` : ''}!`,
        })
    );

    memoized('SpongeBob');
    await delay(300);
    expect(memoized.cache.keys).toEqual([['SpongeBob']]);

    memoized('SpongeBob', 'Patrick');
    await delay(300);
    expect(memoized.cache.keys).toEqual([['SpongeBob', 'Patrick']]);
});

it('should properly match equal keys', async () => {
    const memoized = memoize(() => delay(200));

    memoized(NaN, NaN);
    memoized(NaN, NaN);
});

it('should have a `maxAge` option', async () => {
    const onCacheChange = jest.fn();
    const onExpire = jest.fn();
    const memoized = memoize(() => delay(200), {maxSize: 2, onCacheChange, maxAge: 500, onExpire});

    memoized('SpongeBob');
    expect(memoized.cache.keys).toEqual([['SpongeBob']]);
    expect(memoized.cache).toMatchSnapshot();
    expect(onCacheChange).toHaveBeenCalledTimes(1);

    await delay(300);
    expect(memoized.cache).toMatchSnapshot();
    expect(onCacheChange).toHaveBeenCalledTimes(2);

    memoized('Patrick');
    expect(memoized.cache.keys).toEqual([['Patrick'], ['SpongeBob']]);
    expect(memoized.cache).toMatchSnapshot();
    expect(onCacheChange).toHaveBeenCalledTimes(3);

    await delay(300);
    expect(memoized.cache.keys).toEqual([['Patrick']]);
    expect(memoized.cache).toMatchSnapshot();
    expect(onCacheChange).toHaveBeenCalledTimes(5);

    await delay(300);
    expect(memoized.cache.keys).toEqual([]);
    expect(memoized.cache).toMatchSnapshot();
    expect(onCacheChange).toHaveBeenCalledTimes(6);
    expect(onExpire).toHaveBeenCalledTimes(2);
});

it('should not expire keys if `onExpire` returns false', async () => {
    const onCacheChange = jest.fn();
    const onExpire = jest.fn(() => false);
    const memoized = memoize(() => delay(200), {onCacheChange, maxAge: 500, onExpire});

    memoized('SpongeBob');
    expect(onCacheChange).toHaveBeenCalledTimes(1);
    await delay(300);
    expect(memoized.cache.keys).toEqual([['SpongeBob']]);
    expect(onCacheChange).toHaveBeenCalledTimes(2);

    await delay(300);
    expect(memoized.cache.keys).toEqual([['SpongeBob']]);
    expect(onCacheChange).toHaveBeenCalledTimes(4);
    expect(onExpire).toHaveBeenCalledTimes(1);
});

it('should not expire keys if the key has been hit recently and `updateExpire` is true', async () => {
    const onCacheChange = jest.fn();
    const onExpire = jest.fn();
    const memoized = memoize(() => delay(200), {onCacheChange, maxAge: 500, onExpire, updateExpire: true});

    memoized('SpongeBob');
    expect(onCacheChange).toHaveBeenCalledTimes(1);
    await delay(300);
    expect(memoized.cache.keys).toEqual([['SpongeBob']]);
    expect(onCacheChange).toHaveBeenCalledTimes(2);

    await delay(300);
    expect(memoized.cache.keys).toEqual([['SpongeBob']]);
    expect(onCacheChange).toHaveBeenCalledTimes(2);
    expect(onExpire).toHaveBeenCalledTimes(0);
});

it('should not fail to expire if the key does not exist', async () => {
    const onCacheChange = jest.fn();
    const onExpire = jest.fn();
    const memoized = memoize(() => delay(200), {onCacheChange, maxAge: 500, onExpire});

    memoized('SpongeBob');
    expect(onCacheChange).toHaveBeenCalledTimes(1);
    await delay(300);
    expect(memoized.cache.keys).toEqual([['SpongeBob']]);
    expect(onCacheChange).toHaveBeenCalledTimes(2);

    memoized.cache.keys.pop();
    memoized.cache.values.pop();
    expect(memoized.cache.keys).toEqual([]);

    await delay(300);
    expect(memoized.cache.keys).toEqual([]);
    expect(onCacheChange).toHaveBeenCalledTimes(2);
    expect(onExpire).toHaveBeenCalledTimes(1);
});

it('should export `isMemoized` and `options`', async () => {
    const memoized = memoize(() => delay(200));

    expect(memoized.isMemoized).toBe(true);
    expect(memoized.options).toMatchInlineSnapshot(`
Object {
  "isEqual": [Function],
  "maxAge": Infinity,
  "maxSize": 1,
  "onCacheAdd": [Function],
  "onCacheChange": [Function],
  "onCacheHit": [Function],
  "onExpire": [Function],
  "updateExpire": false,
}
`);
});
