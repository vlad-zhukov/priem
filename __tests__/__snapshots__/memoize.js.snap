// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should export the internal cache 1`] = `
Object {
  "keys": Array [],
  "size": 0,
  "values": Array [],
}
`;

exports[`should export the internal cache 2`] = `
Object {
  "keys": Array [
    Array [
      "world",
    ],
  ],
  "size": 1,
  "values": Array [
    Object {
      "promise": Promise {},
      "reason": null,
      "status": 0,
      "value": null,
    },
  ],
}
`;

exports[`should export the internal cache 3`] = `
Object {
  "keys": Array [
    Array [
      "world",
    ],
  ],
  "size": 1,
  "values": Array [
    Object {
      "promise": Promise {},
      "reason": null,
      "status": 1,
      "value": "Hello world!",
    },
  ],
}
`;

exports[`should have a \`maxAge\` option 1`] = `
Object {
  "keys": Array [
    Array [
      "SpongeBob",
    ],
  ],
  "size": 1,
  "values": Array [
    Object {
      "promise": Promise {},
      "reason": null,
      "status": 0,
      "value": null,
    },
  ],
}
`;

exports[`should have a \`maxAge\` option 2`] = `
Object {
  "keys": Array [
    Array [
      "SpongeBob",
    ],
  ],
  "size": 1,
  "values": Array [
    Object {
      "promise": Promise {},
      "reason": null,
      "status": 1,
      "value": undefined,
    },
  ],
}
`;

exports[`should have a \`maxAge\` option 3`] = `
Object {
  "keys": Array [
    Array [
      "Patrick",
    ],
    Array [
      "SpongeBob",
    ],
  ],
  "size": 2,
  "values": Array [
    Object {
      "promise": Promise {},
      "reason": null,
      "status": 0,
      "value": null,
    },
    Object {
      "promise": Promise {},
      "reason": null,
      "status": 1,
      "value": undefined,
    },
  ],
}
`;

exports[`should have a \`maxAge\` option 4`] = `
Object {
  "keys": Array [
    Array [
      "Patrick",
    ],
  ],
  "size": 1,
  "values": Array [
    Object {
      "promise": Promise {},
      "reason": null,
      "status": 1,
      "value": undefined,
    },
  ],
}
`;

exports[`should have a \`maxAge\` option 5`] = `
Object {
  "keys": Array [],
  "size": 0,
  "values": Array [],
}
`;

exports[`should memoize promises 1`] = `
Object {
  "promise": Promise {},
  "reason": null,
  "status": 0,
  "value": null,
}
`;

exports[`should memoize promises 2`] = `
Object {
  "promise": Promise {},
  "reason": null,
  "status": 1,
  "value": "Hello world!",
}
`;

exports[`should memoize promises 3`] = `
Object {
  "promise": Promise {},
  "reason": null,
  "status": 1,
  "value": "Hello world!",
}
`;

exports[`should memoize promises 4`] = `
Object {
  "promise": Promise {},
  "reason": null,
  "status": 0,
  "value": null,
}
`;

exports[`should memoize promises 5`] = `
Object {
  "promise": Promise {},
  "reason": null,
  "status": 1,
  "value": "Hello SpongeBob!",
}
`;

exports[`should memoize promises 6`] = `
Object {
  "promise": Promise {},
  "reason": null,
  "status": 1,
  "value": "Hello world!",
}
`;

exports[`should not throw on rejected promises 1`] = `
Object {
  "promise": Promise {},
  "reason": null,
  "status": 0,
  "value": null,
}
`;

exports[`should not throw on rejected promises 2`] = `
Object {
  "promise": Promise {},
  "reason": [Error: Hello world!],
  "status": 2,
  "value": null,
}
`;

exports[`should not throw on rejected promises 3`] = `
Object {
  "promise": Promise {},
  "reason": [Error: Hello world!],
  "status": 2,
  "value": null,
}
`;

exports[`should not throw on rejected promises 4`] = `
Object {
  "promise": Promise {},
  "reason": null,
  "status": 0,
  "value": null,
}
`;

exports[`should not throw on rejected promises 5`] = `
Object {
  "promise": Promise {},
  "reason": [Error: Hello SpongeBob!],
  "status": 2,
  "value": null,
}
`;
